### PR TITLE
Extract transfer ownership logic into a reusable service

### PR DESCRIPTION
**Pull request by ChristophWurst**
This will make

* https://github.com/nextcloud/server/pull/17736 possible without code duplication
* https://github.com/nextcloud/guests/pull/208 instant migration possible (quick because no shares) without background job

Tested with a simple guest account.

### CODE REVIEW

File: apps/files/lib/Service/OwnershipTransferService.php [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+			// because the files folder is moved away we need to recreate it
+			$view->mkdir("$sourceUid/files");
+		}
+	}
+
+	private function restoreShares(string $sourceUid,
+								   string $destinationUid,
+								   array $shares,
+								   OutputInterface $output) {
+		$output->writeln("Restoring shares ...");
```
Comment (nickvergessen) [inline apps/files/lib/Service/OwnershipTransferService.php]: `…`

But I know you only moved it, maybe its evevn fine on the console to not use utf8 chars?

File: apps/files/lib/Service/OwnershipTransferService.php [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+	/**
+	 * @param OutputInterface $output
+	 *
+	 * @throws \Exception
+	 */
+	protected function analyse(string $sourceUid,
+							   string $destinationUid,
+							   string $sourcePath,
+							   OutputInterface $output): void {
+		$view = new View();
```
Comment (nickvergessen) [inline apps/files/lib/Service/OwnershipTransferService.php]: inject the view from before?

File: apps/files/lib/Service/OwnershipTransferService.php [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+		$progress->finish();
+		$output->writeln('');
+		return $shares;
+	}
+
+	protected function transferFiles(string $sourceUid,
+									 string $sourcePath,
+									 string $finalTarget,
+									 OutputInterface $output): void {
+		$view = new View();
```
Comment (nickvergessen) [inline apps/files/lib/Service/OwnershipTransferService.php]: same here

File: apps/files/lib/Service/OwnershipTransferService.php (hunk 1 above)
Comment (ChristophWurst) [reply apps/files/lib/Service/OwnershipTransferService.php]: I've had trouble in the past with a tool that printed emojis. That wasn't well supported when running in Docker.

File: apps/files/lib/Service/OwnershipTransferService.php (hunk 2 above)
Comment (ChristophWurst) [reply apps/files/lib/Service/OwnershipTransferService.php]: inject as in create just once and then pass it on or inject in the constructor?

Comment (nickvergessen) [reply apps/files/lib/Service/OwnershipTransferService.php]: The first one

**Review by nickvergessen — APPROVED** [review]
Looks good

**Review by rullzer — APPROVED** [review]
Looks sane!

**Comment by rullzer** [comment]
rebase and autosquash first ;)

**Comment by rullzer** [comment]
And another rebase. But then we can get it in I guess

**Pull request merged**